File: apps/worker/clipforge/publish/rights.py
"""The rights gate.

Ingesting and analysing someone else's video locally is ordinary private use.
*Publishing* a derived clip of third-party material is a different act, and the
legal basis is the operator's to establish — it varies by jurisdiction, by the
source's licence, and by how transformative the clip is.

ClipForge does not pretend that question away and does not answer it either. It
makes the answer **explicit and auditable**: publishing is off by default, and no
clip reaches the publish queue without a recorded basis, an attesting user and a
timestamp.

The gate is enforced in **two independent places** — Firestore rules and this
module — and that redundancy is deliberate. The worker authenticates with the
Admin SDK, which bypasses rules entirely, so rules alone would protect nothing on
the path that actually performs the upload.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime

from clipforge_contracts import Clip, RightsAttestation, RightsBasis

__all__ = [
    "MAX_ATTESTATION_AGE_DAYS",
    "RightsRefusal",
    "RightsViolationError",
    "check_publishable",
    "require_publishable",
]

# Every basis is a positive assertion by a person. There is deliberately no
# "unverified" or "probably fine" value: a placeholder that looks like an
# attestation but asserts nothing is precisely the ambiguity a rights gate exists
# to remove.
ACCEPTED_BASES = frozenset(RightsBasis)

# How stale an attestation may be before it is worth re-confirming. Rights change
# — a licence lapses, permission is withdrawn — and an attestation made a year
# ago is evidence about last year.
MAX_ATTESTATION_AGE_DAYS = 365


@dataclass(frozen=True)
class RightsRefusal:
    """Why a clip may not be published, in terms the operator can act on."""

    code: str
    message: str


class RightsViolationError(RuntimeError):
    """A clip reached the publish path without a usable attestation."""

    # Never retryable. No amount of waiting turns an unattested clip into an
    # attested one; only a person can, and the job should say so rather than
    # burn its attempts discovering that three times.
    retryable = False

    def __init__(self, refusal: RightsRefusal) -> None:
        self.code = refusal.code
        self.refusal = refusal
        super().__init__(refusal.message)


def check_publishable(
    clip: Clip,
    *,
    publishing_enabled: bool,
    now: datetime | None = None,
) -> RightsRefusal | None:
    """Return why this clip may not be published, or ``None`` if it may.

    Returns rather than raises so a caller can list every blocked clip in one
    pass — the UI wants to explain a queue, not stop at the first problem.
    """
    now = now or datetime.now(UTC)

    if not publishing_enabled:
        return RightsRefusal(
            "PUBLISHING_DISABLED",
            "Publishing is disabled. Set CLIPFORGE_PUBLISHING_ENABLED=true once you have "
            "read the rights guidance in the README.",
        )

    if clip.review is not None and clip.review.value != "APPROVED":
        return RightsRefusal(
            "NOT_APPROVED",
            f"This clip is {clip.review.value}. Only an approved clip can be published.",
        )

    attestation = clip.rights
    if attestation is None:
        return RightsRefusal(
            "NO_ATTESTATION",
            "This clip has no rights attestation. Record why you may publish it before "
            "it can enter the publish queue.",
        )

    if attestation.basis not in ACCEPTED_BASES:
        return RightsRefusal(
            "UNKNOWN_BASIS",
            f"{attestation.basis!r} is not a recognised rights basis.",
        )

    if not attestation.attested_by:
        # An attestation with nobody's name against it cannot answer the one
        # question the audit log exists to answer.
        return RightsRefusal(
            "NO_ATTESTOR",
            "The rights attestation records no attesting user, so it cannot be audited.",
        )

    if attestation.attested_at is None:
        return RightsRefusal(
            "NO_ATTESTATION_TIME",
            "The rights attestation has no timestamp, so its age cannot be judged.",
        )

    age_days = (now - attestation.attested_at).days
    if age_days > MAX_ATTESTATION_AGE_DAYS:
        return RightsRefusal(
            "ATTESTATION_STALE",
            f"This attestation is {age_days} days old. Rights change — confirm it still "
            "holds before publishing.",
        )

    if attestation.basis is RightsBasis.FAIR_USE_ASSERTED and not (attestation.note or "").strip():
        # Fair use is a judgement, not a status, and it is the one basis whose
        # reasoning genuinely matters after the fact. An empty note here would
        # make the audit log say "because I said so".
        return RightsRefusal(
            "FAIR_USE_NEEDS_REASONING",
            "A fair-use assertion needs a note explaining the reasoning. It is a judgement, "
            "not a status, and the audit log should record why.",
        )

    return None


def require_publishable(
    clip: Clip, *, publishing_enabled: bool, now: datetime | None = None
) -> RightsAttestation:
    """Assert the clip may be published, and return the attestation that says so.

    Returning the attestation matters: the caller copies it onto the publication
    record, so the reason that justified *this* upload survives a later edit to
    the clip.
    """
    refusal = check_publishable(clip, publishing_enabled=publishing_enabled, now=now)
    if refusal is not None:
        raise RightsViolationError(refusal)
    assert clip.rights is not None  # noqa: S101 - guaranteed by check_publishable
    return clip.rights

File: apps/web/src/app/core/rights.ts
import type { Clip, RightsAttestation, RightsBasis } from '@clipforge/contracts';

/**
 * The rights gate, as the UI sees it.
 *
 * This is the **third** copy of these rules, and the only one that is advisory.
 * The enforcing copies are `firebase/firestore.rules` (which stops a client
 * enqueueing publish work) and `apps/worker/clipforge/publish/rights.py` (which
 * stops the component that actually holds the credentials). Neither of those can
 * explain itself in the interface, and an interface that offered a Publish
 * button and then failed opaquely would be worse than one that never offered it.
 *
 * So: this copy exists to *say why*, immediately and in context. It is not a
 * security boundary and nothing here should ever become one — if this file and
 * the other two disagree, the other two win and this one is the bug.
 *
 * Kept deliberately in step with rights.py, including the refusal codes, so the
 * message a user sees here matches the one a failed job would carry.
 */

export interface RightsRefusal {
  readonly code: string;
  readonly message: string;
}

/** Mirrors `MAX_ATTESTATION_AGE_DAYS` in rights.py. */
export const MAX_ATTESTATION_AGE_DAYS = 365;

/** The choices, in the order they are offered, with the plain-English framing. */
export const RIGHTS_BASES: readonly {
  readonly value: RightsBasis;
  readonly label: string;
  readonly hint: string;
}[] = [
  {
    value: 'OWN_CONTENT',
    label: 'I made this',
    hint: 'You filmed or recorded the source yourself.',
  },
  {
    value: 'LICENSED',
    label: 'I have a licence',
    hint: 'Creative Commons, a stock licence, or a paid licence that permits this use.',
  },
  {
    value: 'PERMISSION_GRANTED',
    label: 'The owner said yes',
    hint: 'Explicit permission from the rights holder. Keep the message.',
  },
  {
    value: 'PUBLIC_DOMAIN',
    label: 'Public domain',
    hint: 'Copyright has expired or was never claimed.',
  },
  {
    value: 'FAIR_USE_ASSERTED',
    label: 'Fair use / fair dealing',
    hint: 'A judgement you are making. Explain your reasoning — it goes in the audit log.',
  },
];

/**
 * Why this clip may not be published, or `null` if it may.
 *
 * Returns rather than throws so a caller can explain a whole queue in one pass.
 */
export function checkPublishable(
  clip: Clip,
  options: { readonly publishingEnabled?: boolean; readonly now?: Date } = {},
): RightsRefusal | null {
  const now = options.now ?? new Date();

  if (clip.review !== 'APPROVED') {
    return {
      code: 'NOT_APPROVED',
      message: `This clip is ${clip.review ?? 'unreviewed'}. Only an approved clip can be published.`,
    };
  }

  const rights = clip.rights as RightsAttestation | null | undefined;
  if (!rights) {
    return {
      code: 'NO_ATTESTATION',
      message: 'Record why you may publish this clip before it can enter the publish queue.',
    };
  }

  if (!RIGHTS_BASES.some((basis) => basis.value === rights.basis)) {
    return { code: 'UNKNOWN_BASIS', message: `“${rights.basis}” is not a recognised basis.` };
  }

  if (!rights.attestedBy) {
    return {
      code: 'NO_ATTESTOR',
      message: 'This attestation records nobody, so it cannot be audited.',
    };
  }

  if (!rights.attestedAt) {
    return {
      code: 'NO_ATTESTATION_TIME',
      message: 'This attestation has no timestamp, so its age cannot be judged.',
    };
  }

  const ageDays = Math.floor((now.getTime() - new Date(rights.attestedAt).getTime()) / 86_400_000);
  if (ageDays > MAX_ATTESTATION_AGE_DAYS) {
    return {
      code: 'ATTESTATION_STALE',
      message: `This attestation is ${ageDays} days old. Rights change — confirm it still holds.`,
    };
  }

  if (rights.basis === 'FAIR_USE_ASSERTED' && !rights.note?.trim()) {
    return {
      code: 'FAIR_USE_NEEDS_REASONING',
      message:
        'A fair-use assertion needs a note explaining the reasoning. It is a judgement, not a status.',
    };
  }

  return null;
}

/** Whether a draft attestation is complete enough to be worth submitting. */
export function draftIsComplete(basis: RightsBasis | null, note: string): boolean {
  if (!basis) return false;
  // The one basis whose reasoning genuinely matters after the fact.
  if (basis === 'FAIR_USE_ASSERTED') return note.trim().length > 0;
  return true;
}
